mongorestore/mongorestore_test.go

```diff
@@ -273,6 +279,256 @@ func generateTestData() error {
+	Options *bson.D
+}
+
+type ByIndexName []bson.D
+
+func (n ByIndexName) Len() int {
+	return len(n)
+}
+
+func (n ByIndexName) Swap(i, j int) {
+	n[i], n[j] = n[j], n[i]
+}
+
+func (n ByIndexName) Less(i, j int) bool {
+	return getIndexName(n[i]) < getIndexName(n[j])
+}
+
+// removeKey removes the given key. Returns the removed value and true if the
+// key was found.
+func removeKey(key string, document *bson.D) (interface{}, bool) {
+	if document == nil {
+		return nil, false
+	}
+	doc := *document
+	for i, elem := range doc {
+		if elem.Key == key {
+			// Remove this key.
+			*document = append(doc[:i], doc[i+1:]...)
+			return elem.Value, true
+		}
+	}
+	return nil, false
+}
+
+// getCollectionDocs returns a slice of all documents in a collection.
+func getCollectionDocs(coll *mongo.Collection) ([]bson.D, error) {
+	c, err := coll.Find(context.Background(), bson.D{}, driverOpt.Find().SetSort(bson.D{{"_id", 1}}))
+	if err != nil {
+		return nil, err
+	}
+
+	defer c.Close(context.Background())
+	var docs []bson.D
+	for c.Next(context.Background()) {
+		var doc bson.D
+		if err = c.Decode(&doc); err != nil {
+			return nil, err
+		}
+
+		docs = append(docs, doc)
+	}
+
+	return docs, nil
+}
+
+func findStringValueByKey(keyName string, document *bson.D) (string, error) {
+	value, err := bsonutil.FindValueByKey(keyName, document)
+	if err != nil {
+		return "", err
+	}
+	str, ok := value.(string)
+	if !ok {
+		return "", fmt.Errorf("field present, but is not a string: %v", value)
+	}
+	return str, nil
+}
+
+func getIndexName(index bson.D) string {
+	name, err := findStringValueByKey("name", &index)
+	if err != nil {
+		return ""
+	}
+	return name
+}
+
+// isViewError checks if err is an "CommandNotSupportedOnView" error.
+func isViewError(err error) bool {
+	e, ok := err.(mongo.CommandError)
+	return ok && e.Code == 166
+}
+
+// bsonDToM converts the given bson.D to bson.M.
+func bsonDToM(d *bson.D) *bson.M {
+	if d == nil {
+		return nil
+	}
+	m := make(bson.M)
+	for _, elem := range *d {
+		m[elem.Key] = elem.Value
+	}
+	return &m
+}
+
+// fullCollectionName returns the full namespace for a collection (dbName.collectionName)
+func fullCollectionName(c *mongo.Collection) string {
+	return fmt.Sprintf("%s.%s", c.Database().Name(), c.Name())
+}
+
+func getIndexDocumentsForCollection(c *mongo.Collection) ([]bson.D, error) {
+	var indexes []bson.D
+	indexesIter, err := c.Indexes().List(context.Background())
+	if err != nil {
+		if isViewError(err) {
+			return indexes, nil
+		}
+		return nil, err
+	}
+	for indexesIter.Next(context.Background()) {
+		var index bson.D
+		if err := indexesIter.Decode(&index); err != nil {
+			return nil, err
+		}
+
+		indexes = append(indexes, index)
+	}
+	sort.Sort(ByIndexName(indexes))
+	return indexes, nil
+}
+
+// AssertBsonDEqualUnordered, asserts two bson.Ds are equal ignoring the
+// order of the top level fields. Useful for comparing indexes.
+func assertBsonDEqualUnordered(expected, actual *bson.D) {
```

Comment:
You might want to recurse on `assertBsonDEqualUnordered` if you really want to make this robust. The collations can be `bson.D`s so it's theoretically possible that could cause issues. Or you could make `bsonDToM` recurse. On the other hand, I think that's pretty unlikely.

Did you run into issues using `assertBsonDEqual()` with index documents? If `assertBsonDEqual()` works, maybe that's just easier to use.